packages/rum/src/viewCollection.ts

```diff
@@ -0,0 +1,171 @@
+import { PerformancePaintTiming } from './rum'
+import { RumSession } from './rumSession'
+import { trackEventCounts } from './trackEventCounts'
+
+export interface View {
+  id: string
+  location: Location
+  measures: ViewMeasures
+  documentVersion: number
+  startTime: number
+  duration: number
+}
+
+export interface ViewMeasures {
+  firstContentfulPaint?: number
+  domInteractive?: number
+  domContentLoaded?: number
+  domComplete?: number
+  loadEventEnd?: number
+  errorCount: number
+  resourceCount: number
+  longTaskCount: number
+  userActionCount: number
+}
+
+const THROTTLE_VIEW_UPDATE_PERIOD = 3000
+
+export function startViewCollection(location: Location, lifeCycle: LifeCycle, session: RumSession) {
+  let currentLocation = { ...location }
+  const startOrigin = 0
+  let currentView = newView(lifeCycle, currentLocation, session, startOrigin)
+
+  // Renew view on history changes
+  trackHistory(() => {
+    if (areDifferentViews(currentLocation, location)) {
+      currentLocation = { ...location }
+      currentView.end()
+      currentView = newView(lifeCycle, currentLocation, session)
+    }
+  })
+
+  // Renew view on session changes
+  lifeCycle.subscribe(LifeCycleEventType.SESSION_RENEWED, () => {
+    currentView.end()
+    currentView = newView(lifeCycle, currentLocation, session)
+  })
+
+  // End the current view on page unload
+  lifeCycle.subscribe(LifeCycleEventType.WILL_UNLOAD, () => {
+    currentView.end()
+  })
+}
+
+interface ViewContext {
+  id: string
+  location: Location
+  sessionId: string | undefined
+}
+
+export let viewContext: ViewContext
+
+function newView(
+  lifeCycle: LifeCycle,
+  location: Location,
+  session: RumSession,
+  startOrigin: number = performance.now()
+) {
+  // Setup initial values
+  const id = generateUUID()
+  let measures: ViewMeasures = {
+    errorCount: 0,
+    longTaskCount: 0,
+    resourceCount: 0,
+    userActionCount: 0,
+  }
+  let documentVersion = 0
+
+  viewContext = { id, location, sessionId: session.getId() }
+
+  // Update the view every time the measures are changing
+  const scheduleViewUpdate = throttle(monitor(updateView), THROTTLE_VIEW_UPDATE_PERIOD, {
+    leading: false,
+  })
+  function updateMeasures(newMeasures: Partial<ViewMeasures>) {
+    measures = { ...measures, ...newMeasures }
+    scheduleViewUpdate()
+  }
+  const { stop: stopTimingsTracking } = trackTimings(lifeCycle, updateMeasures)
+  const { stop: stopEventCountsTracking } = trackEventCounts(lifeCycle, updateMeasures)
+
+  // Initial view update
+  updateView()
+
+  function updateView() {
+    documentVersion += 1
+    lifeCycle.notify(LifeCycleEventType.VIEW_COLLECTED, {
+      documentVersion,
+      id,
+      location,
+      measures,
+      duration: performance.now() - startOrigin,
+      startTime: startOrigin,
+    })
+  }
+
+  return {
+    end() {
+      stopTimingsTracking()
+      stopEventCountsTracking()
+      // Make a final view update
+      updateView()
```

Comment:
is there a test that ensure that this last update is sent with the latest timings / event counts?